warehouse/tuf/services.py

```diff
@@ -0,0 +1,378 @@
+                    threshold=self._request.registry.settings[
+                        f"tuf.{Role.BIN_N.value}.threshold"
+                    ],
+                    keys=self._key_storage_backend.get(Role.BIN_N.value, "private"),
+                    delegation_role=bin_n_name,
+                    path_hash_prefixes=bin_n_hash_prefixes,
+                )
+            )
+
+        metadata_repository.delegate_targets_roles(
+            delegate_roles_payload,
+            self._set_expiration_for_role(Role.TIMESTAMP.value),
+            self._set_expiration_for_role(Role.SNAPSHOT.value),
+        )
+
+    def bump_snapshot(self):
+        # Bumping the Snapshot role involves the following steps:
+        # 1. Initiate Metadata Repository.
+        # 2. Load the Snapshot Role.
+        # 3. Bump Snapshot role (and write to the Storage).
+        # 4. Bump Timestamp role using the new Snapshot Metadata.
+
+        # 1. Metadata Repository.
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # 2. Snapshot role metadata.
+        snapshot_metadata = metadata_repository.load_role(Role.SNAPSHOT.value)
+
+        # 3. Bump Snapshot role metadata.
+        snapshot_metadata = metadata_repository.snapshot_bump_version(
+            snapshot_expires=self._set_expiration_for_role(
+                self._request, Role.SNAPSHOT.value
+            ),
+            store=True,
+        )
+
+        # 4. Bump Snapshot role metadata using Timestamp metadata.
+        metadata_repository.timestamp_bump_version(
+            snapshot_version=snapshot_metadata.signed.version,
+            timestamp_expires=self._set_expiration_for_role(
+                self._request, Role.TIMESTAMP.value
+            ),
+            store=True,
+        )
+
+    def bump_bin_n_roles(self):
+        # Bumping all of the delegated bin roles in the TUF repository involves
+        # the following steps:
+        # 1. Metadata Repository.
+        # 2. Load Snapshot role.
+        # 3. Load BINS role.
+        # 4. For each BIN-N role (hash bin) in BINS, fetch the
+        #    role, bump, write back to the repo and update Snapshot role MetaFile.
+        # 5. Bump Snapshot and write back to repository.
+        # 6. Bump Timestamp role (using updated Snapshot) and write back to
+        #    repository.
+
+        # 1. Metadata Repository
+        metadata_repository = MetadataRepository(
+            self._storage_backend, self._key_storage_backend
+        )
+
+        # 2. Load Snapshot role.
+        snapshot_metadata = metadata_repository.load_role(Role.SNAPSHOT.value)
+
+        # 3 Load BINS role.
+        bins_metadata = metadata_repository.load_role(Role.BINS.value)
+
+        # 4. Fore each delegated hashed bin target role, bump and update Snapshot
+        for role in bins_metadata.signed.delegations.roles.keys():
```

Comment:
We could also iterate over the hash bins generator, which should yield the exact same names as `bins_metadata`. One file less to read might be a minimal performance gain. 🤷